[Hamill] 코드 리뷰 요청

hanurii opened the pull request:
## 처음부터 정렬까지 구현했습니다
- 기본 자료구조 구현
- 미션 구현

hanurii committed:
[Hamill] Feat : 재귀, 순차리스트, 연결리스트, 더미연결리스트 구현

hanurii committed:
Merge branch 'hamill' of https://github.com/Data-Structure-Study/java-datastructure into hamill

hanurii committed:
[Hamill] Feat : 원형연결리스트, 양방향연결리스트, 배열스택, 연결리스트스택, 계산기 구현

hanurii committed:
[Hamill] Feat : 원형큐, 연결리스트큐, 햄버거 시뮬레이션, 덱, 덱 기반 큐 구현

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/basic/BinarySearch.java:
```diff
@@ -0,0 +1,31 @@
+
+  private static int binarySearch(int[] numbers, int number, int first, int last) {
+    while (first <= last) {
+      int mid = (first + last) / 2;
```
이 부분은 int overflow를 조심하셔야해요.
https://endorphin0710.tistory.com/112
여길 참고해보세요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/basic/BinarySearchOperationCount.java:
```diff
@@ -0,0 +1,44 @@
+public class BinarySearchOperationCount {
+
+  private static int binarySearch(int[] numbers, int number, int first, int last) {
+    int mid;
```
이건 약간 윤성우님 책 스타일로 선언을 했네요 ㅋㅋ
저라면 초기화와 동시에 선언하는 스타일을 사용했을 것 같습니다.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/basic/BinarySearchOperationCount.java:
```diff
@@ -0,0 +1,44 @@
+    int opCount = 0;
+
+    while (first <= last) {
+      mid = (first + last) / 2;
```
여기도 리뷰 참고해서 수정하시길 바랄게요!

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/basic/BinarySearchOperationCount.java:
```diff
@@ -0,0 +1,44 @@
+
+  private static int binarySearch(int[] numbers, int number, int first, int last) {
+    int mid;
+    int opCount = 0;
```
제가 윤성우 책에서 안좋아하는 부분이 명확하지 않은 변수명이에요.
약어를 사용하시기 보다는 보다 명확한 변수명을 사용하시는게 좋을 것 같아요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/basic/LinearSearch.java:
```diff
@@ -0,0 +1,26 @@
+public class LinearSearch {
+
+  private static int linearSearch(int[] numbers, int number) {
+    int i;
+    for (i = 0; i < numbers.length; i++) {
```
앗... 아아...

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/basic/LinearSearch.java:
```diff
@@ -0,0 +1,26 @@
+    if (index == -1) {
+      System.out.println("못찾았다");
+    } else {
+      System.out.println((index+1) + "번째 에서 찾았다");
```
```suggestion
      System.out.println((index + 1) + "번째 에서 찾았다");
```

사소한거지만, 이런건 띄어쓰는게 관례입니다.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+package com.codesquad.datastructurestudy.list.arraylist;
+
+public class ArrayList implements List<Integer> {
```
특이하네요 ㅋㅋㅋ 제네릭을 이렇게 구현하신 이유가 있을까요?

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+  private int numOfData;
+
+  public ArrayList() {
+    this.numbers = new int[100];
+    this.numOfData = 0;
```
음 가급적이면 숫자 상수도 private static final로 빼서 사용하는 편이 좋지 않을까 싶어요.
그리고 numOfData는 그냥 제거해도 0으로 초기화 되니 그 점을 염두에 두시는게 좋을 것 같네요.

ksundong replied:
추가로...
저라면 생성자를 
```java
public ArrayList(int initialSize) {
  this.numbers = new int[initialSize];
}
```

이런 식으로 만들고 생성자 오버로딩을 통해서 

```java
public ArrayList() {
  this(INITIAL_SIZE);
}
```

와 같은 식으로 구현해봤을 것 같네요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+    this.numOfData = 0;
+  }
+
+  public ArrayList(int[] numbers, int numOfData) {
+    this.numbers = numbers;
+    this.numOfData = numOfData;
+  }
```
살짝 애매한 생성자 같아요.
이 생성자가 과연 필요할까? 라는 생각도 들고요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+  public void add(int index, Integer element) {
+    if (numOfData > this.numbers.length) {
+      System.out.println("배열이 가득 찼습니다");
+//      throw new ArrayIndexOutOfBoundsException();
```
안쓰는 주석은 제거해주세요~

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+    this.numOfData = numOfData;
+  }
+
+  public void add(int index, Integer element) {
+    if (numOfData > this.numbers.length) {
+      System.out.println("배열이 가득 찼습니다");
+//      throw new ArrayIndexOutOfBoundsException();
+    }
+    this.numbers[index] = element;
+    numOfData++;
+  }
```
이 메서드 확실한가요?
행동하는 것은 replace에 가까운데, 하는 동작에 군더더기가 많이 들어간 느낌입니다.
특히 개수 체크하는 로직은 없어도 될 것 같고, 마찬가지로 `numOfData++;` 도 불필요해보입니다.
그리고 `IndexOutOfBoundException`도 발생가능한 코드입니다.

1. add인데, 만약 0번 위치에 계속 데이터를 집어넣으면? 나중엔 배열이 가득 찼다고 하면서 데이터는 바뀌고, `numOfData`도 계속 증가할겁니다. if문이 종료되고 아래 라인들이 실행될 필요가 없다면 `return`문을 사용해주세요.
2. `numOfData`라는 멤버변수에 `this` 키워드를 추가해주세요. 헷갈릴 수 있습니다.
3. 23번 라인에 `IndexOutOfBoundException`이 발생가능한 코드가 그대로 있습니다. 예외 처리를 추가해주시던지, if 문으로 조건을 달아주세요. (인덱스의 범위를 생각해봅시다.)
4. 23번 라인에 element가 unboxing될 때, 어떻게 될지도 생각해봅시다. `null`이 그대로 들어가게 된다면? 어떤 결과가 벌어질까요?

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+    numOfData++;
+  }
+
+  public void add(Integer element) {
+    if (numOfData > this.numbers.length) {
+      System.out.println("배열이 가득 찼습니다");
+    }
+    this.numbers[numOfData] = element;
+    numOfData++;
+  }
```
여기도 마찬가지입니다. 위의 코멘트를 참고해서 정확한 로직으로 변경해주세요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+    numOfData++;
+  }
+
+  public int count() {
+    return this.numOfData;
+  }
```
메서드명은 관례상 `size()`가 좀 더 적합해보입니다.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+    return this.numOfData;
+  }
+
+  public Integer get(int index) {
+    return this.numbers[index];
+  }
```
`wrapper type`으로 반환하게 되면 `auto boxing`이 일어나기 때문에
여기선 불필요하다고 보입니다.
`Integer`와 `int` 타입은 많은 차이가 있습니다.
특히 메모리 공간, Nullable 등의 차이가 있겠네요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+    return this.numbers[index];
+  }
+
+  public Integer remove(int index) {
+    Integer data = this.numbers[index];
+
+    for (int i = index; i < this.numOfData - 1; i++) {
+      this.numbers[i] = this.numbers[i+1];
+    }
+
+    this.numOfData--;
+    return data;
+  }
```
데이터가 없다면 0을 주겠네요?
뭔가 이상하지 않을까요 ㅎㅎ

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+    Integer data = this.numbers[index];
+
+    for (int i = index; i < this.numOfData - 1; i++) {
+      this.numbers[i] = this.numbers[i+1];
```
```suggestion
      this.numbers[i] = this.numbers[i + 1];
```

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ArrayList.java:
```diff
@@ -0,0 +1,85 @@
+    return data;
+  }
+
+  public int size() {
+    return this.numbers.length;
+  }
```
응? 이거 어떻게보면 맞는 것 같기도 하고 아닌 것 같기도 합니다.
java의 size는 어떻게 동작하나요?

ksundong reviewed (changes requested):
밥먹고 또 리뷰하겠습니다.

몇 몇 변경이 필요한 부분이 있어서 change request 드려요 ㅎㅎ

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/List.java:
```diff
@@ -0,0 +1,27 @@
+
+  int size();
+
+//  int indexOf(T o);
+//  int lastIndexOf(T o);
+
+//  ListIterator listIterator();
+//  ListIterator listIterator(int index);
+
+//  T set(int index, T element);
+
+//  void sort(Comparator c);
+
+//  List subList(int fromIndex, int toIndex);
```
ㅋㅋㅋ 나중에 구현할 사항들인가요?

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ListMain.java:
```diff
@@ -0,0 +1,35 @@
+  public static void main(String[] args) {
+    List<Integer> list = new ArrayList<>();
+    for (int i = 0; i < 9; i++) {
+      list.add(i,i+1);
```
전반적으로 java coding convention이 잘 지켜지지 않는 모습이네요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/ListMain.java:
```diff
@@ -0,0 +1,35 @@
+      }
+    }
+
+    Iterator<Integer> iterator = list1.iterator();
+    while (iterator.hasNext()) {
+      System.out.println(iterator.next());
+    }
```
enhanced for loop를 사용해도 됩니다.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/NameCard.java:
```diff
@@ -0,0 +1,50 @@
+package com.codesquad.datastructurestudy.list.arraylist;
+
+public class NameCard {
+
+  private String[] name;
+  private String[] phone;
```
좀 더 객체지향적인 접근을 해보세요 ㅎㅎ

ksundong replied:
요 코드는 다시 작성하는게 좋아보입니다.
전체적으로 자바로 짠 느낌이 하나도 들지 않았어요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/arraylist/Wallet.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface Wallet {
+
+  // 모든 멤버변수는 public static final. 생략 가능
+  int coin100Num = 0;
+  int bill5000Num = 0;
+
+  // 모든 메서드는 public abstract 이어야함. 생략 가능
```
학습 내용은 좋아보입니다.
근데 보통 interface에서는 멤버변수 선언을 잘 안하는 편이고, 네이밍 컨벤션에 맞게 작성해주시는게 좋을 것 같아요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+    private String name;
+    private Node next;
+
+    public Node() {
+      this.data = 0;
+      this.next = null;
+    }
+
+    public Node(int data) {
+      this.data = data;
+      this.next = null;
+    }
```
생성자에 대해서 좀만 더 공부해보세요. 불필요한 초기화 코드는 제거해주는게 좋습니다.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+  private Node before;
+  private Node cur;
+
+  public void add(int data) {
+    Node newNode = new Node(data);
+
+    if (tail == null) {
+      tail = newNode;
+      newNode.next = newNode;
+    } else {
+      newNode.next = tail.next;
+      tail.next = newNode;
+      tail = newNode;
+    }
+    count++;
+  }
```
```suggestion
  public void add(int data) {
    this.addName(data, null);
  }
```

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+    count++;
+  }
+
+  public void addFront(int data) {
+    Node newNode = new Node(data);
+
+    if (tail == null) {
+      tail = newNode;
+      newNode.next = newNode;
+    } else {
+      newNode.next = tail.next;
+      tail.next = newNode;
+    }
+    count++;
+  }
```
```suggestion
  public void addFront(int data) {
    this.addNameFront(data, null);
  }
```

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  public void addNameFront(int data, String name) {
+    Node newNode = new Node(data);
```
생성자를 하나 더 추가하는 것도 좋아보입니다.
`Node newNode = new Node(data, name);`

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+    count++;
+  }
+
+  public boolean first() {
```
메소드명만 보고 어떤 기능을 하는지 알 수 없네요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  public int remove() {
+    Node deleteNode = cur;
+    int data = deleteNode.data;
+
+    if (deleteNode == tail) {
+      if (tail == tail.next) {
+        tail = null;
+      } else {
+        tail = before;
+      }
+    }
+
+    before.next = cur.next;
+    cur = before;
+    count--;
+    return data;
```
`this` 를 가능한한 붙여주세요.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+    int count = 0;
+    String watcher = null;
+
+    while (true) {
+      if (searchNode.name == name) {
+        break;
+      } else {
+        searchNode = searchNode.next;
+        count++;
+      }
+    }
```
```suggestion
    while (searchNode.name != name) {
      searchNode = searchNode.next;
      count++;
    }
```

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+  public String identifyWatcher(String name, int afterNum) {
+    Node searchNode = new Node();
+    searchNode.next = tail.next;
+    int count = 0;
```
지역 변수와 멤버 변수명이 겹치네요 ㅎㅎ

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+      }
+    }
+
+    for (int i = 0; i < afterNum + count; i++) {
+      if (i != afterNum + count - 1) {
```
`afterNum + count` 중복되는 변수는 따로 빼줍시다!

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/CircularLinkedList.java:
```diff
@@ -0,0 +1,164 @@
+    for (int i = 0; i < afterNum + count; i++) {
+      if (i != afterNum + count - 1) {
+        searchNode = searchNode.next;
+        continue;
```
continue는 필요 없습니다.

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/DDoublyLinkedList.java:
```diff
@@ -0,0 +1,74 @@
+package com.codesquad.datastructurestudy.list.linkedlist.yoon;
+
+public class DDoublyLinkedList {
```
👀

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/DDoublyLinkedList.java:
```diff
@@ -0,0 +1,74 @@
+  private Node tail;
+  private int count;
+
+  public DDoublyLinkedList() {
+    this.dummyNodeHead = new Node(-1);
+    this.dummyNodeTail = new Node(-2);
+
+    this.dummyNodeHead.prev = null;
+    this.dummyNodeHead.next = this.dummyNodeTail;
+
+    this.dummyNodeTail.prev = dummyNodeHead;
+    this.dummyNodeTail.next = null;
+
+    this.head = this.dummyNodeHead;
+    this.tail = this.dummyNodeTail;
+  }
```
생성자가 너무 웅장합니다...

ksundong replied:
굳이 dummyNode 부분을 분리해야 했을까요?

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/DDoublyLinkedList.java:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  public void remove() {
+    if (tail.prev == head) {
+      System.out.println("삭제할 노드가 없습니다");
+    }
```
출력 이후에 아래의 로직이 실행됩니다~

ksundong commented on hamill/src/main/java/com/codesquad/datastructurestudy/list/linkedlist/yoon/DDoublyLinkedList.java:
```diff
@@ -0,0 +1,74 @@
+    return this.count;
+  }
+
+  public static void main(String[] args) {
+
+  }
```
0ㅇ0

ksundong reviewed (changes requested):
1. 의미있는 변수명 사용하기
2. if문으로 예외를 처리할 때, 출력만 하지말고, 종료 포인트 잡아주기
3. 생성자 생성시점에 초기화되는 값은 굳이 불필요하게 안적어도 되면 적지말기
4. 자바 컨벤션 지켜주기

이 밖에도 많은 코멘트가 있습니다.
참고해서 수정해주세요 ㅎㅎ...

hanurii committed:
[Hamill] Feat : 연결리스트 기반 이진트리, 이진트리 순회, 이진 트리 삭제, 수식 트리 구현

hanurii committed:
[Hamill] Feat : 우선순위 큐, 힙 구현

- 힙은 완전이진트리 형태이며, 값에 우선순위에 따라 최대힙, 최소힙으로 불린다.
  - 작은 값이 우선순위가 높으면 최소힙, 큰 값이 우선순위가 높으면 최대힙이라고 한다.
- 우선순위 큐는 들어갈 때는 큐와 다를바 없이 들어가지만 나올 때는 우선순위가 가장 높은 것이 나오는 자료구조다.
  - 배열, 연결리스트, 힙으로 구현할 수 있으며 각각 장단점이 있지만 힙으로 구현했을 때가 가장 우선순위 큐 구현에 적절하다.
